src/traceme.c

```diff
@@ -403,39 +577,57 @@ traceme_start_tracer(struct tracer_conf *tr_conf)
  *
  * Arguments:
  *   tr_ctx: The tracer context returned by traceme_start_tracer.
+ *
+ * Returns 0 on success and -1 on failure.
  */
-void
+int
 traceme_stop_tracer(struct tracer_ctx *tr_ctx) {
 #ifdef TRAVIS
     (void) tr_ctx;
     DEBUG("Travis mode. Not stopping tracing");
+    return -1;
 #else
     DEBUG("stopping tracer");
+
+    int ret = 0;
+
     // Turn off tracer hardware.
     if (ioctl(tr_ctx->perf_fd, PERF_EVENT_IOC_DISABLE, 0) < 0) {
-        err(EXIT_FAILURE, "ioctl");
+        ret = -1;
     }
 
     // Signal poll loop to end.
-    if (close(tr_ctx->loop_done_fds[1]) == -1) {
-        err(EXIT_FAILURE, "close");
+    if (close(tr_ctx->stop_fd_wr) == -1) {
+        ret = -1;
     }
 
     // Wait for poll loop to exit.
     DEBUG("wait for trace thread to exit");
-    if (pthread_join(tr_ctx->tracer_thread, NULL) != 0) {
-        err(EXIT_FAILURE, "pthread_join");
+    void *thr_exit;
+    if (pthread_join(tr_ctx->tracer_thread, &thr_exit) != 0) {
+        ret = -1;
+    }
+    if ((intptr_t) thr_exit != 0) {
```

Comment:
I was sort-of expecting `thr_exit` to be dereferenced, because aren't you just testing whether the pointer you've received is `NULL` or not (which it will never be due to the nature of your calls to `pthread_exit`).

Reply:
`pthread_exit` only allows you to return a void pointer:
```
     void
     pthread_exit(void *value_ptr);
```
but we want to return an int. By using a `intptr_t` we can safely cast our void pointer to the int. It's not really a pointer.

Reply:
Yes, I got that -- but I think you need to do `if (*((intptr_t *) thr_exit) != 0)` or similar.

Reply:
So I think we solved this above.